KG-437 Use structured ModelInfo instead of 'provider:model' string

fazlerahmanejazi opened the pull request:
## Motivation and Context
[KG-437](https://youtrack.jetbrains.com/issue/KG-437/Provide-more-details-for-model-parameters-for-agent-events)
- Swaps the "provider:model" string for a typed ModelInfo data class across agent events.
- Added ModelInfo in agents-utils (no circular deps)
- LLModel.toModelInfo() lives in prompt-llm
- Kept eventString for backward compatibility
- Wired into serialization; tests updated and green

## Breaking Changes
- No breaking changes for consumers using the old string—it's still parsed. New code should prefer ModelInfo.
---

#### Type of the changes
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Tests improvement
- [ ] Refactoring

#### Checklist
- [x] The pull request has a description of the proposed change
- [x] I read the [Contributing Guidelines](https://github.com/JetBrains/koog/blob/main/CONTRIBUTING.md) before opening the pull request
- [x] The pull request uses **`develop`** as the base branch
- [x] Tests for the changes have been added
- [x] All new and existing tests passed

##### Additional steps for pull requests adding a new feature
- [x] An issue describing the proposed change exists
- [x] The pull request includes a link to the issue
- [ ] The change was discussed and approved in the issue
- [ ] Docs have been added / updated

Ololoshechkin committed:
Allow adjusting context window sizes for Ollama dynamically (#883)

kpavlov committed:
Refactor spring boot starters (#886)

Refactor LLM client auto-configuration structure

- Removed `KoogAutoConfiguration` and related legacy configuration
files.
- Introduced modular `LLM` provider-specific auto-configuration classes
(e.g., `OpenAILLMAutoConfiguration`).
- Updated test setup to align with the new modular configuration.
- Enhanced property validation with `enabled` flag support for
finer-grained control.
- Introduced a `String.masked` extension to obfuscate sensitive string
parts. Added test coverage for various input scenarios, including custom
mask characters and edge cases.
- Updated `KoogAutoConfigurationTest` to include scenarios where
providers are explicitly disabled.
- Refactored parameterized tests to use `textBlock = PROVIDERS` with
`@CsvSource`.
- Adjusted `ApplicationContextRunner` configuration logic for clean test
scenarios.
- Introduced `ignoreUnknownFields` to all `ConfigurationProperties` to
improve flexibility with unmapped fields.
- Expanded property documentation for all LLM clients (e.g., OpenAI,
Google, Anthropic) to include detailed descriptions, usage examples, and
retry behavior.
- Improved autowiring safety with `@ConditionalOnBean` for LLM-specific
executors.
- Updated `spring-boot.md` to detail `enabled` flags and their default
behavior for LLM providers.
- Improved clarity by reformatting the supported provider list.
- Added explanations for default base URLs and explicit activation for
non-API-key providers (e.g., Ollama).

A blog post about writing Spring Boot starters is
[here](https://kpavlov.me/blog/spring-boot-starters)

**NB! Tests need a split and some love, but this PR is already big
enough. I will refactor them separately.**

## Motivation and Context

- The Spring Boot configuration in general is not very Spring Boot
idiomatic
- The current Spring Boot starter does not expose LLM Clients. In order
to get the LLM client in the application, developers have to use
reflection to a get a field from the executor 🤯
- All configuration is encapsulated in one file, which makes
extensibility difficult (SRP is violated).
- Defaults are hardcoded in classes instead of being provided in the
resources
- Ollama is enabled by the presence of the base URL, which is
interesting.
- There is no explicit way to disable an LLM provider: no `enabled=true`
property

## Breaking Changes

- AutoConfiguration classes moved to their respective packages
- Ollama provider needs `ai.koog.ollama.enabled=true` to be set
explicitly in order to activate.

---

#### Type of the changes
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or feature that would cause existing
functionality to change)
- [x] Documentation update
- [x] Tests improvement
- [x] Refactoring

#### Checklist
- [x] The pull request has a description of the proposed change
- [ ] I read the [Contributing
Guidelines](https://github.com/JetBrains/koog/blob/main/CONTRIBUTING.md)
before opening the pull request
- [x] The pull request uses **`develop`** as the base branch
- [ ] Tests for the changes have been added
- [ ] All new and existing tests passed

##### Additional steps for pull requests adding a new feature
- [ ] An issue describing the proposed change exists
- [ ] The pull request includes a link to the issue
- [ ] The change was discussed and approved in the issue
- [x] Docs have been added / updated

nicolasf committed:
fix: Google: use maxTokens from params (#734)

Commit:
use structured ModelInfo instead of 'provider:model' string

kpavlov commented on agents/agents-utils/src/commonMain/kotlin/ai/koog/agents/utils/ModelInfo.kt:
```diff
@@ -0,0 +1,56 @@
+     * Backwards compatibility: provides the same string format as current eventString
+     * Format: "provider:model"
+     */
+    public val eventString: String get() = "$provider:$model"
```
Since it's use-case specific, wouldn't it be better to move it to an extension function in the events module? Or let's not mention "events" use case here (rename)

fazlerahmanejazi replied:
This will be deprecated, keeping it for backwards compatibility(improved docs and added deprecation warnings).

kpavlov commented on agents/agents-utils/src/commonTest/kotlin/ai/koog/agents/utils/ModelInfoTest.kt:
```diff
@@ -0,0 +1,73 @@
+
+    @Test
+    fun `test fromString with invalid format`() {
+        val modelInfo = ModelInfo.fromString("invalid-format")
```
Should we create such malformed ModelInfos, or isn't it better to throw a ParseException. We may define a static constant `ModelInfo.UNDEFINED` for undefined usecases

fazlerahmanejazi replied:
Good point! Added ModelInfo.UNDEFINED

kpavlov reviewed (approved):
Thank you, @fazlerahmanejazi 
I have Couple of notes, but in general looks good to me

kpavlov commented:
The ModelInfo might be reused for #633

sdubov commented on agents/agents-utils/src/commonMain/kotlin/ai/koog/agents/utils/ModelInfo.kt:
```diff
@@ -0,0 +1,56 @@
+     * Backwards compatibility: provides the same string format as current eventString
+     * Format: "provider:model"
+     */
+    public val eventString: String get() = "$provider:$model"
```
Let's deprecate this as well. I think we need to drop it in the next release in favour of a new format string: `provider/model_id` below.

sdubov commented on agents/agents-utils/src/commonMain/kotlin/ai/koog/agents/utils/ModelInfo.kt:
```diff
@@ -0,0 +1,56 @@
+     * Human-readable display name for the model
+     * Falls back to "provider/model" if displayName is not provided
+     */
+    public val humanReadableName: String get() = displayName ?: "$provider/$model"
```
Maybe name it `modelIdentifierName` or `modelIdentifierString`?

sdubov commented on agents/agents-core/src/commonMain/kotlin/ai/koog/agents/core/feature/model/events/llmCallEvents.kt:
```diff
@@ -40,19 +49,27 @@ public data class LLMCallStartingEvent(
+     * @deprecated Use model.eventString instead
+     */
+    @Deprecated("Use model.eventString instead", ReplaceWith("model.eventString"))
+    public val modelString: String get() = model.eventString
```
Maybe also add a deprecated constructor to keep backward compatibility:
```
/**
     * @deprecated Use constructor with model parameter of type [ModelInfo]: 
     *             LLMCallStartingEvent(runId, prompt, model, tools, eventId, timestamp)
     */
    @Deprecated(
        message = "Please use constructor with model parameter of type [ModelInfo]: LLMCallStartingEvent(runId, prompt, model, tools, eventId, timestamp)",
        replaceWith = ReplaceWith("LLMCallStartingEvent(runId, prompt, model, tools, eventId, timestamp)")
    )
    public constructor(
        runId: String,
        prompt: Prompt,
        model: String,
        tools: List<String>,
        eventId: String = LLMCallStartingEvent::class.simpleName!!,
        timestamp: Long = Clock.System.now().toEpochMilliseconds()
    ) : this(runId, prompt, ModelInfo.fromString(model), tools, eventId, timestamp)
```

sdubov reviewed (changes requested):
Thank you for your contribution. I have several comments about naming and deprecation. It would be nice to fix them. Other than that LGTM!

Commit:
add UNDEFINED constant, deprecate eventString, add backwards compatibility constructors

sdubov reviewed (approved):
Thank you for fixes. Approve!

kpavlov reviewed (approved):
Thank you, @fazlerahmanejazi

Commit:
resolve merge conflicts

Commit:
fix build

Commit:
updated replaceWith and docs

fazlerahmanejazi commented on agents/agents-core/src/commonMain/kotlin/ai/koog/agents/core/feature/model/events/llmCallEvents.kt:
```diff
@@ -16,18 +17,43 @@ import kotlinx.serialization.Serializable
+     */
+    @Deprecated(
+        message = "Please use constructor with model parameter of type [ModelInfo]: LLMCallStartingEvent(runId, prompt, model, tools, eventId, timestamp)",
+        replaceWith = ReplaceWith("LLMCallStartingEvent(runId, prompt, model, tools, eventId, timestamp)")
```
done, both #905 and #893 would need a workflow!

Commit:
fix: update Qodana linter to compatible version 2025.2

Commit:
resolve flaky PersistenceRunsTwiceTest by using correct rollback strategy

Commit:
fix: wait for server to be ready before client connection in DebuggerTest

Commit:
fix lint

Commit:
Resolve merge conflicts by preferring develop branch version

Commit:
Merge remote-tracking branch 'upstream/develop' into kg-437-agent-events-modelinfo

Commit:
Restore KoogAutoConfigurationTest.kt to upstream develop version

- Fix unintended changes from previous merge conflict resolution
- Restore RetryConfig.DEFAULT and createApplicationContextRunner() method
- Ensure test file matches upstream develop branch exactly

Commit:
Restore Spring Boot auto-configuration files to upstream develop version

- Fix unintended changes from previous merge conflict resolution
- Restore all LLM auto-configuration classes to upstream develop
- Restore Spring Boot properties files to upstream develop
- Restore Spring Boot documentation to upstream develop
- Restore qodana.yaml to upstream develop
- Ensure only intended agent events and model info changes remain

Commit:
Fix debugger test timeout issues

- Remove unnecessary messageProcessor.initialize() calls from tests
- These calls were causing conflicts with existing server startup logic
- Tests now pass successfully without timeout errors
- Related to ModelInfo changes in kg-437-agent-events-modelinfo branch

Commit:
Update Qodana linter to version 2025.2

- Sync with upstream develop branch changes
- Fixes compatibility with Qodana CLI 2025.2.1
- Resolves token requirement issue from version 2025.1

fazlerahmanejazi commented:
@sdubov Thanks for fixing the flaky tests in DebuggerTest, I believe this could be merged as well!

kpavlov committed:
Merge branch 'develop' into kg-437-agent-events-modelinfo

Commit:
compile issue after merge

fazlerahmanejazi committed:
Merge branch 'develop' into kg-437-agent-events-modelinfo

Commit:
lint fix

Commit:
stabilize debugger and tracing remote writer tests
